test/integration/replicaset/replicaset_test.go

```diff
@@ -272,246 +280,595 @@ func TestAdoption(t *testing.T) {
+		_, err = podClient.Update(newPod)
+		return err
 	}); err != nil {
-		t.Fatal(err)
+		t.Fatalf("Failed to update .OwnerReferences for pod %s: %v", pod.Name, err)
 	}
-	pod2, err = podClient.Get(pod2.Name, metav1.GetOptions{})
+
+	if err := wait.PollImmediate(interval, timeout, func() (bool, error) {
+		newPod, err := podClient.Get(pod.Name, metav1.GetOptions{})
+		if err != nil {
+			return false, err
+		}
+		if metav1.GetControllerOf(newPod) != nil {
+			return true, nil
+		}
+		return false, nil
+	}); err != nil {
+		t.Fatalf("Failed to verify ControllerRef for the pod %s is not nil: %v", pod.Name, err)
+	}
+
+	rs, err := c.ExtensionsV1beta1().ReplicaSets(ns).Get(rsName, metav1.GetOptions{})
 	if err != nil {
-		t.Fatalf("Failed to get pod2: %v", err)
+		t.Fatalf("failed to get replicaset %s: %v", rsName, err)
+	}
+	controllerRef := metav1.GetControllerOf(pod)
+	if controllerRef.UID != rs.UID {
+		t.Fatalf("RS owner of the pod %s has a different UID: Expected %v, got %v", pod.Name, rs.UID, controllerRef.UID)
 	}
-	if len(pod2.OwnerReferences) != 0 {
-		t.Fatalf("ownerReferences of pod2 is not cleared, got %#v", pod2.OwnerReferences)
+	ownerReferenceNum := len(pod.GetOwnerReferences())
+	if ownerReferenceNum != expectedOwnerReferenceNum {
+		t.Fatalf("Unexpected number of owner references for pod %s: Expected %d, got %d", pod.Name, expectedOwnerReferenceNum, ownerReferenceNum)
 	}
-	close(stopCh)
 }
 
-func TestUpdateLabelToBeAdopted(t *testing.T) {
-	// We have pod1, pod2 and rs. rs.spec.replicas=1. At first rs.Selector
-	// matches pod1 only; change pod2's labels to be matching. Verify the RS
-	// controller adopts pod2 and delete one of them, so there is only 1 pod
-	// left.
-	s, closeFn, rm, informers, clientSet := rmSetup(t)
+func TestGeneralPodAdoption(t *testing.T) {
+	s, closeFn, rm, informers, c := rmSetup(t)
 	defer closeFn()
-	ns := framework.CreateTestingNamespace("rs-update-label-to-be-adopted", s, t)
+	ns := framework.CreateTestingNamespace("test-general-pod-adoption", s, t)
 	defer framework.DeleteTestingNamespace(ns, s, t)
+	labelMap := map[string]string{"foo": "bar"}
 	rs := newRS("rs", ns.Name, 1)
-	// let rs's selector only matches pod1
-	rs.Spec.Selector.MatchLabels["uniqueKey"] = "1"
-	rs.Spec.Template.Labels["uniqueKey"] = "1"
-	pod1 := newMatchingPod("pod1", ns.Name)
-	pod1.Labels["uniqueKey"] = "1"
-	pod2 := newMatchingPod("pod2", ns.Name)
-	pod2.Labels["uniqueKey"] = "2"
-	createRSsPods(t, clientSet, []*v1beta1.ReplicaSet{rs}, []*v1.Pod{pod1, pod2}, ns.Name)
+	rs.Spec.Selector.MatchLabels = labelMap
+	rs.Spec.Template.Labels = labelMap
+	rss, _ := createRSsPods(t, c, []*v1beta1.ReplicaSet{rs}, []*v1.Pod{}, ns.Name)
+	rs = rss[0]
 
 	stopCh := make(chan struct{})
 	informers.Start(stopCh)
 	go rm.Run(5, stopCh)
-	waitRSStable(t, clientSet, rs, ns.Name)
+	defer close(stopCh)
+	waitRSStable(t, c, rs, ns.Name)
+
+	podSelector := labels.Set{"foo": "bar"}.AsSelector()
```

Comment:
This code chunk should be refactored as a helper function as it appears multiple times throughout the codebase.